subprojects/docs/src/docs/userguide/five_things.adoc

```diff
@@ -0,0 +1,126 @@
+
+Tasks themselves consist of:
+
+ * Actions — pieces of work that do something, like copy files or compile source
+ * Inputs — values, files and directories that the actions use or operate on
+ * Outputs — files and directories that the actions modify or generate
+
+In fact, all of the above are optional depending on what the task needs to do. Some tasks — such as the <<base_plugin#sec:base_tasks,standard lifecycle tasks>> — don't even have any actions. They simply aggregate multiple tasks together as a convenience.
+
+NOTE: You choose which task to run. Save time by specifying the task that does what you need, but no more than that. If you just want to run the unit tests, choose the task that does that — typically `test`. If you want to package an application, most builds have an `assemble` task for that.
+
+One way to customize a build is to add extra steps to it. This is easy to do in Gradle: define your own tasks and insert them into the task graph via the <<tutorial_using_tasks#sec:task_dependencies,task dependency mechanism>>. You can have your tasks depend on any other task or configure existing tasks to depend on yours. The main restriction is that you cannot create any circular dependencies within the task graph.
+
+One last thing: Gradle's <<more_about_tasks#sec:up_to_date_checks,incremental build>> support is robust and reliable, so keep your builds running fast by avoiding the `clean` task unless you actually do want to perform a clean. 
+
+== 3. Gradle has distinct build phases
+
+It's important to understand that Gradle evaluates and executes build scripts in three phases:
+
+ 1. Initialization
+ 2. Configuration
+ 3. Execution
+
+In well-designed build scripts, the vast majority of what you see is evaluated during the configuration phase. Even so, some elements are likely to be evaluated during the execution phase, particularly the contents of `doLast {}` and `doFirst {}` blocks.
+
+NOTE: Gradle's build phases are not like Maven's phases. Maven uses its phases to divide the build execution into multiple stages. They serve a similar role to Gradle's task graph, although less flexibly.
+
+When you are authoring or modifying a build, it's important to recognize which bits are evaluated during which phases. Otherwise you will end up with a build that doesn't work as expected. For example, if you set a property within the execution phase, any part of the build that reads that property during the configuration phase won't see the new value.
+
+Consider the following (contrived) example that attempts to initialise a `Copy` task from a properties file. You might try this if the properties file is generated by another task. The problem is that `into` must be set during the configuration phase, otherwise the task won't run. Setting the value within a `doFirst {}` block fails as it happens too late.
+
+[source,groovy]
+.Configuring a task within an action (execution phase) won't work
+----
+task copyFiles(type: Copy) {
+    doFirst {
```

Comment:
❌ This will actually work just fine. It only becomes problematic when you try to make changes that would have affected the task graph. This is not such a change though.